test_functests_gevent_server.py

```diff
@@ -0,0 +1,178 @@
+# -*- coding: utf-8 -*-
+from __future__ import unicode_literals
+import gevent_server
+import socket
+import pytest
+import time
+from multiprocessing import Process
+
+
+addr = ("127.0.0.1", 8000)
+_CRLF = b'\r\n'
+
+
+# yield fixtures are demons
+
+# We used to have a yield fixture here and in the server.py tests
+# which would start up the server and keep it going, I thought, until
+# the end of its scope. Since these server fixtures were scoped to
+# module, we believed they would terminate at the end of the module.
+# The theory seemed to hold true through the end of Step 3, since we
+# only ever started one server throughout the entire testing process.
+# Once we created the gevent server, there were within the test suite
+# two different server creation fixtures, both scoped to module. We
+# falsely believed that each of these fixtures would terminate at the
+# end of the module. In practice, it seems that a yield fixture doesn't
+# terminate until the end of the entire testing session, regardless
+# of defined scope.
+
+# The solution, seen below, is to use just a regular fixture with
+# a process-terminating finalizer. The scope behaves properly,
+# and autouse also still works.
+
+
+@pytest.fixture(scope='module', autouse=True)
+def gevent_server_setup(request):
+    process = Process(target=gevent_server.run_gevent_server)
+    process.daemon = True
+    process.start()
+    time.sleep(0.1)
+
+    def cleanup():
+        process.terminate()
+
+    request.addfinalizer(cleanup)
+    return process
+
+
+@pytest.fixture(scope='function')
+def client_setup():
+    client = socket.socket(
+        socket.AF_INET, socket.SOCK_STREAM, socket.IPPROTO_IP
+    )
+    client.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
+    client.connect(addr)
+    return client
+
+
+def verify_response(response):
+    assert 2 * _CRLF in response
+    head_and_body = response.split((2 * _CRLF), 1)
+    head_chunk = head_and_body[0].split(_CRLF)
+    first_line = head_chunk[0].split()
+    assert first_line[0] == b"HTTP/1.1"
+    assert first_line[1].isdigit()
+    assert first_line[2] is not None
```

Comment:
given the tests you have below, though, it might be nice to have this method hand back some data like a first line and a dict of headers.  Then you could use that to improve the assertions you make in the tests themselves.